### ISSUE TITLE
"Trigger Suggest" and "Quick Fix..." commands, code navigation not responding

### ISSUE THREAD

**Issue by an-boxbrite** [microsoft/vscode-java-pack#948]
**Environment:**

- VS Code Version: 1.65.x (also tested in 1.64.2)
- Extension Version: 0.22.1 (`vscode-java` 1.4.0, also tested with 1.3.0)
- JDK Type & Version: Amazon Corretto JDK 11.0.11.9.1
- OS Type & Version: Windows

**Step to Reproduce:**
Unknown, seemingly random.

However, restarting VS Code does seem to reset the issue but only temporarily, and the amount of time before the issue resurfaces varies, sometimes as short as a couple minutes, other times it lasts quite a while (hours, days).

**Expected Behavior:**
- Triggering commands "Trigger Suggest" and "Quick Fix..." (i.e. Ctrl+Space, Alt+Enter) should consistently cause intellisense popup to be populated with suggestions/fixes
- All code navigation (Ctrl+click, etc.)/"Go to ___"/"Find all references" commands should work
- Tests can be executed

**Actual Behavior:**
- Triggering mentioned commands manually using keybindings causes intellisense popup to appear, but says "Loading..." indefinitely. The following two entries are added to the LS log upon the Ctrl+Space ("Trigger Suggest") command:
  ```
  !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:08:00.387
  !MESSAGE >> workspace/executeCommand java.intellicode.completion
  
  !ENTRY com.microsoft.jdtls.intellicode.core 1 0 2022-03-28 10:08:00.401
  !MESSAGE Completion request completed
  ```
  - Upon certain context-sensitive keystrokes, such as `.` (period after a reference, to access a member) and `@` to add an annotation, suggestions do populate as expected. The following log entries are added upon typing `.` (period) after a reference:
    <details>
      <summary>View log entries</summary>
    <p>[looks like only the first five entries are relevant, but figured I'd include all just in case]<p>
    
    <pre>!ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.078
    !MESSAGE >> document/didChange
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.085
    !MESSAGE >> workspace/executeCommand java.intellicode.completion
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.090
    !MESSAGE Reconciled 1. Took 1 ms
    
    !ENTRY com.microsoft.jdtls.intellicode.core 1 0 2022-03-28 10:10:58.101
    !MESSAGE Completion request completed
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.126
    !MESSAGE >> workspace/executeCommand java.intellicode.completion.resolve
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.265
    !MESSAGE >> document/foldingRange
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.330
    !MESSAGE >> document/codeAction
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.344
    !MESSAGE >> workspace/executeCommand vscode.java.resolveMainMethod
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.362
    !MESSAGE >> workspace/executeCommand vscode.java.resolveMainMethod
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.369
    !MESSAGE >> textDocument/semanticTokens/full
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.593
    !MESSAGE begin problem for /DataFileServiceImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.593
    !MESSAGE 0 problems reported for /DataFileServiceImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.607
    !MESSAGE begin problem for /DataFileIterator.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.608
    !MESSAGE 0 problems reported for /DataFileIterator.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.629
    !MESSAGE begin problem for /DataFileFactoryImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.629
    !MESSAGE 2 problems reported for /DataFileFactoryImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.881
    !MESSAGE begin problem for /BinaryDataFileWriterImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.882
    !MESSAGE 0 problems reported for /BinaryDataFileWriterImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.895
    !MESSAGE begin problem for /InternalDataFileFactory.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.896
    !MESSAGE 0 problems reported for /InternalDataFileFactory.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:58.896
    !MESSAGE Validated 5. Took 403 ms
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:10:59.028
    !MESSAGE >> document/codeAction
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:00.075
    !MESSAGE >> document/didSave
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:00.389
    !MESSAGE >> workspace/didChangeWatchedFiles 
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:00.395
    !MESSAGE Reconciled 1. Took 0 ms
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.117
    !MESSAGE begin problem for /DataFileServiceImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.117
    !MESSAGE 0 problems reported for /DataFileServiceImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.128
    !MESSAGE begin problem for /DataFileIterator.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.128
    !MESSAGE 0 problems reported for /DataFileIterator.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.147
    !MESSAGE begin problem for /DataFileFactoryImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.147
    !MESSAGE 2 problems reported for /DataFileFactoryImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.219
    !MESSAGE begin problem for /BinaryDataFileWriterImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.219
    !MESSAGE 0 problems reported for /BinaryDataFileWriterImpl.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.234
    !MESSAGE begin problem for /InternalDataFileFactory.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.234
    !MESSAGE 0 problems reported for /InternalDataFileFactory.java
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.234
    !MESSAGE Validated 5. Took 387 ms
    
    !ENTRY org.eclipse.jdt.ls.core 1 0 2022-03-28 10:11:01.412
    !MESSAGE >> document/codeAction
    </pre></details>

  - Other keystrokes, such as letters, which would normally suggest identifiers/members/imports/templates/etc. don't even trigger the "Loading..." popup.
- Some code navigation, "Go to ___" and "Find all ___" commands not responding
  - Not working: "Go to Definition", "Go to References", "Find All References"
  - Working: "Go to Implementation", "Go to Super Implementation", "Go to Type Definition", "Find All Implementations"
  - For those not working:
    - the loading bar beneath tabs appears and moves indefinitely
    - tried triggering using Ctrl+click, keystrokes and context menu commands
- A couple times these have resulted in significant, sustained CPU usage
- `vscode-java-test` plugin appears to have crashed (not sure if this is related)
  - extension is greyed-out in the extension list, cannot be enabled/disabled
  - no "Run" icon beside test cases in code editor
  - test panel not populated

**Related Issues:** 
- microsoft/vscode#145309 (originally [commented on this issue](https://github.com/microsoft/vscode/issues/145309#issuecomment-1077766014))
- microsoft/TypeScript#48517

**Comment by Eskibear** [comment]
Thank you for the detailed description. 

For completion part, because you have IntelliCode extension installed, completion is delegated to IntelliCode. But the log you pasted looks fine. **Can you share a list of your installed extension?** It's possible that other extension also registered a completion provider which stuck there, in this case vscode will keep waiting results from all completion provider and show "loading...".
And in the mean time, you can try with IntelliCode extension disabled, to test if it's caused by IntelliCode.

For other part, I'll invite other extension maintainer to help. cc @jdneo

It would be much helpful if you can attach the full log, or even share a sample project.

**Comment by an-boxbrite** [comment]
Thanks for your response.

I've tried disabling IntelliCode and will let you know what happens. Here is the full list of extensions I have installed:

<details>
<summary>Full extension list</summary>

<pre>
42Crunch.vscode-openapi
atlassian.atlascode
DavidAnson.vscode-markdownlint
dgileadi.java-decompiler
eamodio.gitlens
EditorConfig.EditorConfig
esbenp.prettier-vscode
GabrielBB.vscode-lombok
mechatroner.rainbow-csv
ms-azuretools.vscode-docker
ms-dotnettools.csharp
ms-vscode-remote.remote-wsl
ms-vsliveshare.vsliveshare
ms-vsliveshare.vsliveshare-audio
ms-vsliveshare.vsliveshare-pack
Pivotal.vscode-boot-dev-pack
Pivotal.vscode-spring-boot
PKief.material-icon-theme
redhat.java
redhat.vscode-xml
redhat.vscode-yaml
ryanluker.vscode-coverage-gutters
SonarSource.sonarlint-vscode
VisualStudioExptTeam.vscodeintellicode
vscjava.vscode-gradle
vscjava.vscode-java-debug
vscjava.vscode-java-dependency
vscjava.vscode-java-pack
vscjava.vscode-java-test
vscjava.vscode-spring-boot-dashboard
vscjava.vscode-spring-initializr
zxh404.vscode-proto3
</pre>
</details>

Is it the Java Language Server Log that you're interested in (as I included in the original post), or a different one?

**Comment by Eskibear** [comment]
yes, server log. see https://github.com/redhat-developer/vscode-java/wiki/Troubleshooting#enable-logging for more information. 

If you turn on "verbose" logging, you'll see response of each request. (Note that log file will be much larger and possibly separated into serveral files). You can also take a look by your own before submitting the full log, E.g. for a completion request, if the response contains correct results, but it's still showing "loading...", that means Java language server is working well, but something else is preventing vscode from showing the results.

**Comment by Eskibear** [comment]
> both commands appear to result in correct request/response messages, so it seems the issue is not with the Java language server.

Agree. So it's likely that one of other extensions listed above registered malformed providers. 

To find out the culprit, what I suggest is:
- First, to confirm Java language server is 100% working. Run "Developer: Reload with Extensions Disabled" from command palette, and manually enable "Language Support for Java" extension only, see if the problem is gone. 
- Second, to find out the culprit. You can do either of the following:
  - Manually enable other extensions one by one, reload window, see when the problem is back.
  - Or run "Help: Start Extension Bisect", which might be more convenient. (but if vscode happens to disbale Language support for Java, you will not be able to tell whether the problem is still there).

**Comment by an-boxbrite** [comment]
I'm not really sure how to prove it, but I'm suspicious the Test Runner for Java (`vscjava.vscode-java-test`) is the extension causing problems. I tried disabling all extensions and then enabled only a minimal set to be able to continue my work:

```
redhat.java
vscjava.vscode-java-test
vscjava.vscode-java-debug
SonarSource.sonarlint-vscode
GabrielBB.vscode-lombok
PKief.material-icon-theme
```

`vscode-java-test` was the last one I enabled, though it also requires `vscode-java-debug`, so it was enabled at the same time. I hadn't encountered the issue prior to enabling it, though of course with this issue it's hard to prove its absence. However, since enabling it I have encountered the issue. As I [noted in the original post](https://github.com/microsoft/vscode-java-pack/issues/948#issue-1183543555), I also observed issues with the `vscode-java-test` extension around the same time, though I have not seen those exact symptoms again recently. So two questions then:

1. What does it look like if an extension "crashes"? Do the symptoms I describe in the original post related to `vscode-java-test` make sense?
2. Does the `vscode-java-test` extension register provider(s) that may be causing issues? Not sure how best to determine this.

**Comment by jdneo** [comment]
test runner won't register anything related with completion/go to definition.

It would help if you can provide a sample project with some repro steps to let us investigate

**Comment by Eskibear** [comment]
> vscode-java-test plugin appears to have crashed (not sure if this is related)
> - extension is greyed-out in the extension list, cannot be enabled/disabled
> - no "Run" icon beside test cases in code editor
> - test panel not populated

It seems that vscode-java-test extension is not activated at all. 
- Command `Developer: Show Running Extensions` can help to list running extensions. 
- View -> Output -> select "Log(Extension Host)", see if there's any error related to vscode-java-test.
- If no clues found in above steps, let's assume vscode-java-test extension is broken during installation. You can try totally remove it, and reinstall. 
  - Close vscode, remove directory `C:\users\<username>\.vscode\extensions\vscjava.vscode-java-test-<version>`, open vscode ,install vscode-java-test extension from marketplace.

**Comment by an-boxbrite** [comment]
> It would help if you can provide a sample project with some repro steps to let us investigate

Unfortunately, I have not been able to intentionally reproduce the issue, it just happens sporadically, so I don't know what would be relevant for a sample project. The actual project I'm working on cannot be shared. It's a Gradle multi-project build using Lombok, if that helps.

> It seems that vscode-java-test extension is not activated at all.

Right, I suppose that would have been the case, though it had in the same session previously been active, hence why I wondered about it having crashed. That being said, I have not run into this situation recently, just a couple of times early on, so it's very possible it was unrelated. I'll keep your suggested steps in mind in case it happens again.

**Comment by Cristian-Serrano** [comment]
The same thing has happened to me.

I'm using a java project. If I open VS Code with the terminal in the folder containing src, target, and pom.xml I get the same problem.

But if I open VS Code with the terminal from the src folder, the navigation works, for example the shortcut Ctrl + Click.

In your case, have you found the cause of your problem?
